some changes to make this more generic than just MEG

choldgraf opened the pull request:
After spending some time trying to get MNE-BIDS to work on my ECoG data, I found that there were a few places in there that made assumptions about MEG. I tried to relax those assumptions by allowing for a "kind" keyword where you can specify the type of data we're dealing with. Currently that only supports `['meg', 'ieeg']`. I also changes some namings etc so that the word "meg" isn't the only thing that shows up :-)

jasmainak commented:
Thanks a lot! We are missing tests, so maybe we can merge @teonbrooks PR https://github.com/mne-tools/mne-bids/pull/27 before merging this one

jasmainak commented:
okay @choldgraf you need to rebase and add tests :)

choldgraf commented:
ok, pushed the changes that @jasmainak @agramfort and I talked about this morning. This modularizes the code a bit more, and also exposes some functions for creating bids-compatible filenames + folder hierarchies that might be useful for users. The tests are woefully missing, but I want to give folks a chance to comment before this diverges too much with some other PR out there.

Note also makes it possible for `raw_to_bids` to accept a Raw object as input. All of the fif reading in logic is now in a separate function, and if the input raw file is already an instance of `mne.io.Raw`, it doesn't try to do the reading in stuff. So now it'll work *both* with strings as a path, and as Raw objects.

jasmainak commented:
cc @teonbrooks

agramfort commented on examples/create_bids_folder.py:
```diff
@@ -0,0 +1,54 @@
+###############################################################################
+# We'll import the relevant functions from the utils module
+
+from mne_bids.utils import create_folders, filename_bids
```
I have exposing modules called utils as name does not mean anything... 

if create_folders is a very common function then it should be in root module.

agramfort commented on mne_bids/io.py:
```diff
@@ -0,0 +1,37 @@
+                                  preload=False, verbose=verbose)
+
+    # CTF systems
+    elif ext == '.ds':
```
else Exception ....

agramfort commented on mne_bids/meg_bids.py:
```diff
@@ -246,95 +254,97 @@ def raw_to_bids(subject_id, session_id, run, task, raw_fname, output_path,
+    elif isinstance(raw_file, BaseRaw):
+        # Only parse the filename for the extension
+        raw = raw_file
+        raw_filename = raw.filenames[0]
```
I doubt a RawArray will contain filenames attribute

choldgraf replied:
good point - though this could also support a file that has been read into memory from a file. I'll add another check and assume that if `filenames` isn't there, it's a `.fif` file

agramfort commented on mne_bids/utils.py:
```diff
@@ -0,0 +1,126 @@
+    Returns
+    -------
+    filename : str
+        The BIDS filename you wish to create.
```
can you add a docstring example?

jasmainak replied:
I like this function :-)

agramfort commented on tests/test_utils.py:
```diff
@@ -0,0 +1,14 @@
+from mne_bids.utils import _check_types, create_types, filename_prefix
+from collections import OrderedDict
```
import collections first

choldgraf commented:
comments addressed - I think the main thing this needs now is tests.

I also need to make sure this PR still functions as I thought, because I rebased and this resulted in a bunch of conflicts that were introduced w/ this PR from #32 

I wonder if, given the state of tests currently, it'll be better to make sure tests pass now, then do more robust tests later, given that this touches a lot of the codebase and we'll get a lot of rebase conflicts otherwise

choldgraf commented:
I'm trying to get tests to pass but keep getting the "Failed an Initial Quick Test" validator error...not sure why :-/

choldgraf commented:
just to confirm - @jasmainak we _expect_ travis to pass here, right?

agramfort commented:
Yes it should pass if you rebadged on master

jasmainak commented:
Your filenames are incorrect if you’re failing initial test. Also, the
validator cannot handle iEEG files yet

On Sun 11 Mar 2018 at 13:16, Alexandre Gramfort <notifications@github.com>
wrote:

> Yes it should pass if you rebadged on master
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-bids/pull/28#issuecomment-372110881>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APHiotTJP_ldBM9HCwFd0aoX4_9Xmq63ks5tdRV_gaJpZM4ScQhw>
> .
>
-- 
Sent from my iPhone

choldgraf commented:
hmmm, I can't figure out what is wrong here...this is the tree I'm seeing:

```
/tmp/tmp_mne_tempdir_r1uauk9o/
├── dataset_description.json
└── sub-01
    └── ses-01
        ├── meg
        │   ├── sub-01_task-testing_run-01_channel.tsv
        │   ├── sub-01_task-testing_run-01_events.tsv
        │   └── sub-01_task-testing_run-01_meg.fif
        ├── sub-01_ses-01_fid.json
        ├── sub-01_ses-01_meg.json
        └── sub-01_ses-01_scans.tsv
```
this seems reasonable to me, no? @jasmainak what does the tree look like when you test this?

jasmainak commented:
I think the validator expects a `*.json` file in the `meg/` folder to pass the quick test but maybe that's not the best assumption. I could change that if you want

jasmainak commented:
Hmm ... looking at the MEG specification I think there should indeed be a *.json in the meg/ folder. No?

choldgraf commented:
ahh maybe it's that my `meg.json` file is one folder too high in the hierarchy. It's annoying that the validator doesn't say *which* file it thinks it should find, or which has an error

choldgraf commented:
@jasmainak does this require one to have the MEG fork of the bids-validator?

choldgraf commented:
I ask because the tests are failing for me on _master_

jasmainak commented:
Yep, you need the fork of the MEG validator as it is not yet merged into master. We'll update travis once it is merged into master.

jasmainak commented on mne_bids/utils.py:
```diff
@@ -0,0 +1,137 @@
+        The task for a item. Corresponds to "task".
+    acquisition: str | None
+        The acquisition parameters for the item. Corresponds to "acq".
+    run : str | None
```
shouldn't this be an `int`?

jasmainak commented on examples/create_bids_folder.py:
```diff
@@ -0,0 +1,54 @@
+# pieces of metadata, ensuring that they are in the correct order in the
+# final file path. Omitted keys will not be included in the file path.
+
+my_name = filename_bids(subject='test', session='two', task='mytask', suffix='data.csv')
```
how about `make_bids_fname` or `make_bids_filename`?

jasmainak commented on examples/create_bids_folder.py:
```diff
@@ -0,0 +1,54 @@
+print(path_folder)
+
+# Note that passing `create=True` will create the folder hierarchy, ignoring
+# errors if the folder already exists.
```
I feel `create` is kind of a confusing argument. Do you mean to use `overwrite`?

jasmainak commented on mne_bids/io.py:
```diff
@@ -0,0 +1,40 @@
+    elif ext == '.ds':
+        raw = io.read_raw_ctf(raw_fname)
+    else:
+        raise ValueError("Raw file name extension must be one of {}\nGot {}".format(ALLOWED_EXTENSIONS, ext))
```
you need a pep8 checker ;-)

jasmainak commented on mne_bids/meg_bids.py:
```diff
@@ -189,34 +188,37 @@ def _meg_json(raw, task, manufacturer, fname, verbose):
-                'TriggerChannelCount': n_stimchan,
-                }
-    json_output = json.dumps(meg_json, indent=4, sort_keys=True)
+    chs_json = OrderedDict([
```
I find the name `chs_json` a bit of a misnomer ...

jasmainak commented on mne_bids/meg_bids.py:
```diff
@@ -267,99 +273,103 @@ def raw_to_bids(subject_id, session_id, run, task, raw_fname, output_path,
+    coordsystem_fname = op.join(data_path, filename_bids(subject=subject_id, session=session_id, suffix='coordsystem.json'))
+    data_meta_fname = op.join(data_path, filename_bids(subject=subject_id, session=session_id, suffix='%s.json' % kind))
+
+    raw_file_bids = op.join(data_path, filename_bids(subject=subject_id, session=session_id, task=task, run=run, suffix='%s%s' % (kind, ext)))
```
pep 8 :-)

jasmainak commented on mne_bids/tests/test_utils.py:
```diff
@@ -0,0 +1,12 @@
+from collections import OrderedDict
+from mne_bids.utils import _check_types, create_folders, filename_bids
```
have you added a test for `create_folders`?

jasmainak commented on mne_bids/tests/test_utils.py:
```diff
@@ -0,0 +1,12 @@
+from mne_bids.utils import _check_types, create_folders, filename_bids
+
+# All keys work
+PREFIX_DATA = OrderedDict(subject='one', session='two', task='three',
```
why do we need ordered dict here?

jasmainak commented on mne_bids/tests/test_utils.py:
```diff
@@ -0,0 +1,12 @@
+from collections import OrderedDict
```
author name and license missing :-)

jasmainak commented on mne_bids/meg_bids.py:
```diff
@@ -116,7 +111,7 @@ def _scans_tsv(raw, raw_fname, fname, verbose):
             meas_date).strftime('%Y-%m-%dT%H:%M:%S')
 
-    df = pd.DataFrame({'filename': ['%s' % raw_fname],
+    df = pd.DataFrame({'filename': ['%s' % raw_file],
```
why change this? Don't we use `fname` as the convention for filename everywhere in MNE?

choldgraf commented:
ok I think I addressed all of these comments! also added some new tests. Now just need to figure out how to get the validator installed on my machine so that I can make sure the tests actually pass :-P

jasmainak commented:
lol just install nodejs and follow the commands we use in travis :-)

jasmainak commented on mne_bids/tests/test_utils.py:
```diff
@@ -0,0 +1,31 @@
+
+def test_make_filenames():
+    # All keys work
+    PREFIX_DATA = dict(subject='one', session='two', task='three',
```
normally capital letter is reserved for global variables, no?

teonbrooks commented on mne_bids/io.py:
```diff
@@ -0,0 +1,43 @@
+def _parse_ext(raw_fname):
+    """Split a filename into its name and extension."""
+    fname, ext = os.path.splitext(raw_fname)
+    # BTi data is the only file format that does have a file extension
```
that's my carryover error. "...that does not have a file extension"

teonbrooks commented on mne_bids/meg_bids.py:
```diff
@@ -31,26 +35,18 @@
     map_chs = defaultdict(lambda: 'OTHER')
-    map_chs.update(grad='MEGGRAD', mag='MEGMAG', stim='TRIG', eeg='EEG',
-                   eog='EOG', ecg='ECG', misc='MISC', ref_meg='REFMEG')
+    map_chs = dict(grad='MEGGRAD', mag='MEGMAG', stim='TRIG', eeg='EEG',
```
this should use `maps_chs.update` cf `map_desc.update` below to use the `defaultdict`

teonbrooks commented on mne_bids/utils.py:
```diff
@@ -0,0 +1,212 @@
+    return filename
+
+
+def make_bids_folders(subject, session=None, kind=None, root=None,
```
should we include an overwrite in case the some or all the folders exist?

jasmainak replied:
+1

teonbrooks commented on mne_bids/meg_bids.py:
```diff
@@ -267,99 +246,110 @@ def raw_to_bids(subject_id, session_id, run, task, raw_fname, output_path,
+            _, ext = _parse_ext(raw.filenames[0])
+        else:
+            ext = '.fif'
+    data_path = make_bids_folders(subject=subject_id, session=session_id,
```
are you just getting the path? if so, shouldn't `make_dirs=False`?

choldgraf replied:
in this case I believe it's actually creating the directories

teonbrooks replied:
oh ok, I guess this doesn't make the session folders until below

teonbrooks commented on mne_bids/meg_bids.py:
```diff
@@ -267,99 +246,110 @@ def raw_to_bids(subject_id, session_id, run, task, raw_fname, output_path,
+                          verbose)
+
+    make_dataset_description(output_path, name=" ",
+                             bids_version="1.0.2 (draft)", verbose=verbose)
```
should we make bids_version a global variable and hoist it up. I feel having defined in the function will leave to it being forgotten with updates to the codebase

jasmainak replied:
Agreed !

jasmainak commented on mne_bids/utils.py:
```diff
@@ -0,0 +1,212 @@
+    fname = os.path.join(path, 'dataset_description.json')
+    description = OrderedDict([('Name', name),
+                               ('BIDSVersion', bids_version),
+                               ('License', license),
```
this appears to be a reserved keyword

jasmainak commented on mne_bids/utils.py:
```diff
@@ -0,0 +1,212 @@
+        A path to a folder where the description will be created.
+    name : str | None
+        The name of this BIDS dataset.
+    bids_version : str | None
```
not quite sure this should be exposed to users

choldgraf replied:
good point

jasmainak commented on mne_bids/utils.py:
```diff
@@ -0,0 +1,212 @@
+    return path
+
+
+def make_dataset_description(path, name=None, bids_version=None, license=None,
```
we should just define `bids_version` somewhere and make sure we produce files compatible with that BIDS version. It will be a mess to support different versions of BIDS

jasmainak commented on mne_bids/__init__.py:
```diff
@@ -1,2 +1,3 @@
 from .meg_bids import raw_to_bids  # noqa
-from .meg_bids import _mkdir_p  # noqa
+from .utils import (make_bids_folders, make_bids_filename,
```
should `make_bids_filename` be exposed to the user? I was thinking it is more of a convenience private function called from `make_bids_folders`. When do you expect users to actually call it?

choldgraf replied:
I think so - I find it a convenience when creating other files that the mne-bids function doesn't capture (e.g. if I needed to copy over stim files etc)

Either way, the public API for mne-bids is really simple already, so I don't see the harm in it (though I agree we should be thoughtful about what API we expose to people)

choldgraf commented:
ok, I think that I've got all comments addressed, and I *think* that all meaningful tests should pass. **with one exception**: I know that two tests for `bids-validator` will fail...these are for the two `iEEG` fields in the channels `.json` file (it's erroring saying there can't be extra fields in this file).

Fixing this error will require patching `bids-validator`, which has a bit longer timeline than this PR. I think folks should clone this PR to their machine to test there (after commenting out the two `iEEG` field names in the JSON code) to confirm that the tests pass.

What do you all think?

jasmainak commented on mne_bids/config.py:
```diff
@@ -0,0 +1,3 @@
+"""Configuration values for MNE-BIDS."""
+
+BIDS_VERSION = "1.0.2 (draft)"
```
necessary in new file?

jasmainak replied:
or let it be ...

choldgraf replied:
I couldn't figure out a way to import this from one of the other modules without creating a circular import :-/

jasmainak replied:
ah yes, makes sense

jasmainak commented on mne_bids/tests/test_utils.py:
```diff
@@ -0,0 +1,31 @@
+    assert make_bids_filename(subject='one', task='three', suffix='hi.csv') == 'sub-one_task-three_hi.csv'
+
+
+def test_make_folders():
```
missing docstring

jasmainak commented on mne_bids/tests/test_utils.py:
```diff
@@ -0,0 +1,31 @@
+from mne_bids.utils import _check_types, make_bids_folders, make_bids_filename
+import os
+
+def test_make_filenames():
```
missing docstring

jasmainak commented:
why don't we hack the test for now by reading in the json and writing it back without the two fields and making sure the test passes? :-)

jasmainak commented:
Looks close to merge for me!

any reviews @agramfort @teonbrooks ?

jasmainak commented:
do the examples run on your box without any error?

agramfort commented:
I will look when you made travis green again :)

teonbrooks commented:
> why don't we hack the test for now by reading in the json and writing it back without the two fields and making sure the test passes? :-)

@jasmainak, I was lazy and thinking the exact same thing!

jasmainak commented on mne_bids/meg_bids.py:
```diff
@@ -267,99 +247,114 @@ def raw_to_bids(subject_id, session_id, run, task, raw_fname, output_path,
+        _coordsystem_json(raw, unit, orient, manufacturer, coordsystem_fname,
+                          verbose)
+
+    make_dataset_description(output_path, name=" ",
```
why is this public? I think we should really be mindful of what we make public

choldgraf replied:
In this case, because I saw that people in the `bids-examples` folder were including example scripts etc to create this file, and thought that it's a common-enough operation that we may as well create a helper function for it. In my mind, you could use MNE-BIDS to do the initial raw conversion, and then the `utils` functions to do any tweaking that's needed

jasmainak replied:
hmm ... but there is only one `dataset_description` in the folder. You don't think it's easier to edit the  text file than write a function to do it for you? :-P

jasmainak replied:
I think we should just put `n/a` whatever we can't fill in. Then the user can simply grep for `n/a` and fill up those fields. 

Maybe you would need functions to do that. But let's not overfit. For this PR, we should focus on just iEEG and separating raw i/o from folder creation. One step at a time :-)

choldgraf replied:
that works too - though I anticipate using mne-bids to convert data that's not in `.fif` format, so in my case I'll know about the hidden function but other users in a similar situation (e.g. anybody with ieeg data most likely) won't know about this

choldgraf replied:
re: n/a, I think this should be solved a different way. some of the fields actually can't have strings, (e.g. they expect arrays) so we'd need to hard-code that in there. I think instead we can prototype something like having JSON templates of the proper form in a subsequent PR.

choldgraf commented:
ok I think the latest push should solve our travis problem - it removes the ieeg-specific fields before running the validator on the JSON file. I also removed the "make_bids_dataset" from the main module.

I think that a nice compromise here would be to have something like a `templates` folder within `mne-bids` (or maybe the `bids-starter-kit`) that contained those empty JSON files with all possible fields. Basically something to make it easier for people to wrap their heads around all the kinds of information they *could* put into the JSON files would be useful.

also @jasmainak you owe me a beer the next time we see each other, after I had to decipher that brutal error message in bids-validator ;-)

jasmainak commented:
@choldgraf beer it will be in Berkeley this June :-) But how come you were coding and not drinking beer on st. patrick's day? :-P Sorry about the brutal error message. It's fixed now.

Yes, we could perhaps use something like `mne.externals.tempita`, but another PR :-) Can you make Travis green so we can merge this PR? And then we merge Teon's PR.

jasmainak commented on mne_bids/__init__.py:
```diff
@@ -1,2 +1,3 @@
 from .meg_bids import raw_to_bids  # noqa
-from .meg_bids import _mkdir_p  # noqa
+from .utils import (make_bids_folders, make_bids_filename)
+from .config import BIDS_VERSION
```
missing EOF?

jasmainak commented on mne_bids/__init__.py:
```diff
@@ -1,2 +1,3 @@
 from .meg_bids import raw_to_bids  # noqa
-from .meg_bids import _mkdir_p  # noqa
+from .utils import (make_bids_folders, make_bids_filename)
```
do you need parantheses?

jasmainak commented:
@choldgraf when I run the example, I get the error:

```py
In [1]: %run examples/convert_mne_sample.py
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/home/mainak/Desktop/projects/github_repos/mne-bids/examples/convert_mne_sample.py in <module>()
     39             raw_fname=raw_fname, events_fname=events_fname,
     40             output_path=output_path, event_id=event_id,
---> 41             overwrite=True)

TypeError: raw_to_bids() got an unexpected keyword argument 'raw_fname'
```

choldgraf commented:
ok some small fixes made! let's see if travis is happy now.

@jasmainak good catch re: the example, I updated those scripts. The reason I didn't catch it is that the example requires downloading an ~8GB file, which IMO is way, way, way too big for a simple example demonstrating the functionality of the package, so I didn't run it :-)

choldgraf commented:
half of travis seems to be happy, the 3.6 build isn't and I'm not sure why:

https://travis-ci.org/mne-tools/mne-bids/jobs/354937938#L836

It's passing fine on my machine...

jasmainak commented on mne_bids/meg_bids.py:
```diff
@@ -267,99 +249,114 @@ def raw_to_bids(subject_id, session_id, run, task, raw_fname, output_path,
+        _, ext = _parse_ext(raw_file)
+    elif isinstance(raw_file, BaseRaw):
+        # Only parse the filename for the extension
+        # Assume that if no filename attr exists, it's a fif file.
```
why is that? Shouldn't you throw an error?

jasmainak replied:
I would avoid magic

choldgraf replied:
if you create a Raw object from `RawArray`, there is no associated filename, and since we'll be saving it to `fif` that's why we default to `fif` here

jasmainak commented:
@choldgraf it's the 2.7 build which isn't working and I think there is some flaw somewhere in the logic of parsing extensions because of which you are getting this error. I could reproduce locally on my box

jasmainak commented on mne_bids/io.py:
```diff
@@ -0,0 +1,43 @@
+    """Split a filename into its name and extension."""
+    fname, ext = os.path.splitext(raw_fname)
+    # BTi data is the only file format that does not have a file extension
+    if ext == '':
```
we should throw a warning here. Or some log message

choldgraf replied:
I can try to do this, though this code isn't an addition by me (I only repackaged it into a hidden function) and I'm not 100% sure of the specific use-case there...does BTi data just have no extension whatsoever?

choldgraf replied:
ahh the 2.7 problem is a string type check

choldgraf commented:
ok I added in a print statement for the warning that @jasmainak mentioned...we should (in a future PR, I think) convert all of the "prints" to proper log warnings. Other than that, travis seems to be happy now :-)

jasmainak commented:
@choldgraf I get the following error:

```py
In [1]: %run examples/convert_mne_sample.py
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/home/mainak/Desktop/projects/github_repos/mne-bids/examples/convert_mne_sample.py in <module>()
     39             raw_file=raw_file, events_fname=events_fname,
     40             output_path=output_path, event_id=event_id,
---> 41             overwrite=True)

TypeError: raw_to_bids() got an unexpected keyword argument 'events_fname'
```

jasmainak commented:
> The reason I didn't catch it is that the example requires downloading an ~8GB file, which IMO is way, way, way too big for a simple example demonstrating the functionality of the package, and I didn't have the space on my machine or time so I didn't run it :-)

I agree it's too big. We need to use smaller datasets to demonstrate it but ds000117 is quite interesting as it is a proper group study, so you can demo how to loop over subjects and runs using the existing API. Perhaps we could fix this issue in the future :-)

choldgraf commented:
latest push should fix the events path error...(sorry that I can't test the examples myself...maybe we should built a little sphinx-gallery site so that the examples get run w/ the tests

jasmainak commented on mne_bids/tests/test_meg_bids.py:
```diff
@@ -51,9 +55,10 @@ def test_fif():
+                session_id=session_id, raw_file=raw_fname2,
+                events_data=events_fname, output_path=output_path,
                 event_id=event_id, overwrite=True)
+    _strip_ieeg_from_json(output_path)
```
Perhaps I am missing something but why do we need to do this if there is already an argument 'kind' which defaults to 'meg'? I think since the iEEG specification is not yet part of BIDS main specification there should be an option for the user to leave out the iEEG specific fields. What do you think @choldgraf ?

jasmainak commented:
@choldgraf see my last comment. Otherwise LGTM

jasmainak commented:
@choldgraf 

again this error:

```py
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
/home/mainak/Desktop/projects/github_repos/mne-bids/examples/create_bids_folder.py in <module>()
     20 # We'll import the relevant functions from the utils module
     21 
---> 22 from mne_bids import create_folders, filename_bids
     23 
     24 ###############################################################################

ImportError: cannot import name create_folders
```

why can't you test the examples on your box? you don't have mne sample data? I can test the ds000117 for you but the rest are quick to download, no?

choldgraf commented:
Yeah it's just the sample data, I need to clear out space on my hard drive
for the 8gb download

Re:the iEEG stripping, those values will be in the JSON regardless of
whether there are any iEEG channels. They'll just be 0.

On Sun, Mar 18, 2018, 11:36 PM Mainak Jas <notifications@github.com> wrote:

> @choldgraf <https://github.com/choldgraf>
>
> again this error:
>
> ---------------------------------------------------------------------------ImportError                               Traceback (most recent call last)/home/mainak/Desktop/projects/github_repos/mne-bids/examples/create_bids_folder.py in <module>()
>      20 # We'll import the relevant functions from the utils module
>      21 ---> 22 from mne_bids import create_folders, filename_bids
>      23
>      24 ###############################################################################
> ImportError: cannot import name create_folders
>
> why can't you test the examples on your box? you don't have mne sample
> data? I can test the ds000117 for you but the rest are quick to download,
> no?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-bids/pull/28#issuecomment-374067033>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABwSHU9aOqaQP90rAeBxetv0n3WQHfHCks5tfu9jgaJpZM4ScQhw>
> .
>

jasmainak commented:
> Re:the iEEG stripping, those values will be in the JSON regardless of whether there are any iEEG channels. They'll just be 0.

not quite sure I follow ... but I am just saying stripping is probably not the right solution since iEEG is not part of the BIDS specification yet

choldgraf commented:
@jasmainak earlier in this PR you suggested:

> why don't we hack the test for now by reading in the json and writing it back without the two fields and making sure the test passes? :-)

isn't this exactly what this function is doing? The idea is to delete the function within the tests once the bids-validator has been updated...

jasmainak commented:
yes, correct. Sorry for changing my mind about this, but what I am suggesting now is a slightly different approach. We should do something like:

```py
# it should produce json fields as per latest BIDS draft and test should pass
raw_to_bids(kind='meg', ...)
# it should test produce json fields as per BIDS draft + BEP010 extension 
# and tests should fail
raw_to_bids(kind='ieeg', ...) 
```

This way, users can still continue to produce jsons which pass the validator without having to hack anything. You can probably just move your calls to the stripping function from the tests to the last line of `raw_to_bids` and that would do the job. Something like:

```py
if kind == 'meg':
   _strip_ieeg_from_json(output_path)
```

does that sound reasonable to you?

teonbrooks commented:
@choldgraf let me know when you want a review

jona-sassenhagen commented:
What is the degree of EEG support? I see it shows up at one line in the code. Will it support EEG out of the box?

jasmainak commented:
well, the EEG extension of BIDS is not yet merged into BIDS. But there is some support because MEG recordings are sometimes multimodal

I would say PR is welcome as long as you don't break the BIDS validator :-)

jasmainak commented:
@choldgraf is the PR ready from your end?

choldgraf commented:
I still need to implement the last suggestion you made, but my computer got borked and I had to reinstall everything so it's taking a bit to get back to baseline

jasmainak commented:
okay no worries!

choldgraf commented:
ok the latest push _should_ make MEG pass the validator without needing to do fancy stripping etc. If travis passes then I am ready to merge

choldgraf commented:
arg travis is being sooo slow

jasmainak commented:
@choldgraf you have flake8 issues :)

choldgraf commented:
@jasmainak travis is happy!

jasmainak commented:
@agramfort and @teonbrooks, happy with the changes?

On Wed 28 Mar 2018 at 19:26, Chris Holdgraf <notifications@github.com>
wrote:

> @jasmainak <https://github.com/jasmainak> travis is happy!
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-bids/pull/28#issuecomment-376968425>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APHiojrbQEBd3E3GGr4jGjS-c1mCpO_Uks5ti8fBgaJpZM4ScQhw>
> .
>
-- 
Sent from my iPhone

agramfort commented:
thx @choldgraf

choldgraf commented:
![](https://media.giphy.com/media/l0MYH8Q83CXvKzXyM/giphy.gif)

jasmainak commented:
Bravo !

agramfort commented:
what worries me the most is that you have time to find these things on the
web...

choldgraf commented:
![](https://media.giphy.com/media/EsmlrgWNx5v0Y/giphy.gif)

teonbrooks commented:
thanks @choldgraf!
